File: augur-engine/src/publish/social-queue.ts
import { mkdir, writeFile, readdir, readFile, rename } from 'node:fs/promises';
import { join } from 'node:path';
import type { BrandConfig, Prediction, SocialPlatform, SocialQueueEntry } from '../config/types.js';

/** Schedule social posts for a prediction (staggered 2h apart per platform) */
export async function queueSocialPosts(
  prediction: Prediction,
  captions: Record<SocialPlatform, string>,
  brand: BrandConfig,
  siteDir: string,
): Promise<void> {
  const pendingDir = join(siteDir, '_data', 'social', 'pending');
  await mkdir(pendingDir, { recursive: true });

  const baseTime = new Date();
  let offsetHours = 0;

  for (const platform of brand.socialTargets) {
    const caption = captions[platform];
    if (!caption) continue;

    const scheduledAt = new Date(baseTime.getTime() + offsetHours * 60 * 60 * 1000);

    const entry: SocialQueueEntry = {
      brand: prediction.brand,
      horizon: prediction.horizon,
      dateKey: prediction.dateKey,
      platform,
      scheduledAt: scheduledAt.toISOString(),
      caption,
      imagePath: prediction.imagePaths?.[0] ?? '',
      createdAt: new Date().toISOString(),
      postUrl: null,
      retryCount: 0,
      error: null,
      postedAt: null,
    };

    const filename = `${prediction.brand}-${prediction.dateKey}-${platform}.json`;
    await writeFile(join(pendingDir, filename), JSON.stringify(entry, null, 2), 'utf-8');
    console.log(`[social] queued: ${filename} (scheduled ${scheduledAt.toISOString()})`);

    offsetHours += 2;
  }
}

/** Read all pending social posts */
export async function readPendingPosts(siteDir: string): Promise<Array<{ path: string; entry: SocialQueueEntry }>> {
  const pendingDir = join(siteDir, '_data', 'social', 'pending');
  const files = await readdir(pendingDir).catch(() => [] as string[]);

  const posts: Array<{ path: string; entry: SocialQueueEntry }> = [];

  for (const file of files) {
    if (!file.endsWith('.json')) continue;
    const filePath = join(pendingDir, file);
    const raw = await readFile(filePath, 'utf-8');
    const entry = JSON.parse(raw) as SocialQueueEntry;
    posts.push({ path: filePath, entry });
  }

  return posts.filter((p) => new Date(p.entry.scheduledAt) <= new Date());
}

/** Move a post from pending to posted (success) or failed (error) */
export async function movePost(
  filePath: string,
  entry: SocialQueueEntry,
  status: 'posted' | 'failed',
  siteDir: string,
  postUrl?: string,
  error?: string,
): Promise<void> {
  const targetDir = join(siteDir, '_data', 'social', status);
  await mkdir(targetDir, { recursive: true });

  if (status === 'posted') {
    entry.postUrl = postUrl ?? null;
    entry.postedAt = new Date().toISOString();
  } else {
    entry.error = error ?? 'unknown error';
    entry.retryCount += 1;
  }

  const filename = filePath.split('/').pop()!;
  const targetPath = join(targetDir, filename);
  await writeFile(targetPath, JSON.stringify(entry, null, 2), 'utf-8');

  // Remove from pending
  const { unlink } = await import('node:fs/promises');
  await unlink(filePath);

  console.log(`[social] ${status}: ${filename}`);
}

File: augur-engine/src/publish/social_queue.py
"""JSON file-based social posting queue."""

from __future__ import annotations

import json
import logging
from dataclasses import asdict
from datetime import datetime, timedelta, timezone
from pathlib import Path

from ..config.types import BrandConfig, Prediction, SocialQueueEntry

log = logging.getLogger("augur.social")


def queue_social_posts(
    prediction: Prediction,
    captions: dict[str, str],
    brand: BrandConfig,
    site_dir: str,
) -> None:
    """Schedule social posts for a prediction (staggered 2h apart per platform)."""
    pending_dir = Path(site_dir) / "_data" / "social" / "pending"
    pending_dir.mkdir(parents=True, exist_ok=True)

    base_time = datetime.now(timezone.utc)

    for i, platform in enumerate(brand.social_targets):
        caption = captions.get(platform)
        if not caption:
            continue

        scheduled_at = base_time + timedelta(hours=i * 2)

        entry = SocialQueueEntry(
            brand=prediction.brand,
            horizon=prediction.horizon,
            date_key=prediction.date_key,
            platform=platform,
            scheduled_at=scheduled_at.isoformat(),
            caption=caption,
            image_path=prediction.image_paths[0] if prediction.image_paths else "",
            created_at=datetime.now(timezone.utc).isoformat(),
        )

        filename = f"{prediction.brand}-{prediction.date_key}-{platform}.json"
        (pending_dir / filename).write_text(
            json.dumps(asdict(entry), indent=2), encoding="utf-8"
        )
        log.info("queued: %s (scheduled %s)", filename, scheduled_at.isoformat())


def read_pending_posts(
    site_dir: str,
) -> list[tuple[str, SocialQueueEntry]]:
    """Read all pending social posts that are due."""
    pending_dir = Path(site_dir) / "_data" / "social" / "pending"
    if not pending_dir.exists():
        return []

    now = datetime.now(timezone.utc)
    results: list[tuple[str, SocialQueueEntry]] = []

    for f in pending_dir.glob("*.json"):
        data = json.loads(f.read_text(encoding="utf-8"))
        entry = SocialQueueEntry(**data)
        if datetime.fromisoformat(entry.scheduled_at) <= now:
            results.append((str(f), entry))

    return results


def move_post(
    file_path: str,
    entry: SocialQueueEntry,
    status: str,
    site_dir: str,
    post_url: str | None = None,
    error: str | None = None,
) -> None:
    """Move a post from pending to posted (success) or failed (error)."""
    target_dir = Path(site_dir) / "_data" / "social" / status
    target_dir.mkdir(parents=True, exist_ok=True)

    if status == "posted":
        entry.post_url = post_url
        entry.posted_at = datetime.now(timezone.utc).isoformat()
    else:
        entry.error = error or "unknown error"
        entry.retry_count += 1

    filename = Path(file_path).name
    (target_dir / filename).write_text(
        json.dumps(asdict(entry), indent=2), encoding="utf-8"
    )

    # Remove from pending
    Path(file_path).unlink(missing_ok=True)
    log.info("%s: %s", status, filename)
